[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec

wanchaol opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #92126
* #90735
* #90734
* #90733
* #90732
* #91785
* __->__ #90731

This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

wanchaol committed:
[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec

This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

XilunWu reviewed (approved):
Overall LGTM. Just 1 nit and some questions. Thx Wanchao!

wanchaol commented on test/distributed/_tensor/test_device_mesh.py:
```diff
@@ -143,6 +143,17 @@ def test_device_mesh_nd(self):
+        mesh_tensor_2d = torch.arange(8).reshape(4, 2)
+        mesh = DeviceMesh(self.device_type, mesh_tensor_2d)
+        mesh2 = DeviceMesh(self.device_type, mesh_tensor_2d)
+        self.assertEqual(hash(mesh), hash(mesh2))
```
yeah good catch! Actually hash should respect equality so in this case mesh and mesh2 should not hash to the same value, let me update

wanchaol commented on test/distributed/_tensor/test_common_rules.py:
```diff
@@ -143,9 +143,7 @@ def test_einop_linearity(self):
-        mm_func_schema = parse_schema(
-            "aten::mm(Tensor self, Tensor mat2) -> Tensor"
-        )
+        mm_func_schema = parse_schema("aten::mm(Tensor self, Tensor mat2) -> Tensor")
```
Yep we should, some format discrepency from the two code base, let me see if we can make the default formatting applied

XilunWu commented on torch/distributed/_tensor/device_mesh.py:
```diff
@@ -105,7 +105,7 @@ def __init__(
         self.device_type = device_type
         self.mesh = (
-            mesh.detach()
+            mesh
```
then doesn't the original `mesh.detach()` remove `mesh` from the graph? Why should we change it to `mesh` instead?

wanchaol replied:
I think detach was there because I want to avoid user modifying the mesh tensor. Removed it because of only hashing the mesh tensor, but since we should force equality, mesh + id should be hashed, so I added the detach() call back :)

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]

fduwjj commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -270,6 +293,17 @@ def __post_init__(self) -> None:
             self.ndim = len(self.shape)
 
+    def __hash__(self) -> int:
+        return hash((self.mesh, tuple(self.placements), self.shape))
```
Do we also want to include "ndim"?

wanchaol replied:
probably no, `ndim` is inferred from shape, so it represents the same thing and should hash to the same. Also we might want to get rid of ndim and use shape uniformly.

fduwjj reviewed (approved):
Thanks for pushing this to prod.

wanchaol committed:
Update on "[dtensor][1/N] add __hash__ to device_mesh and dtensor_spec"


This PR adds __hash__ to device_mesh and dtensor_spec to allow
things like dict indexing

[ghstack-poisoned]
